### PR TITLE
Feature http xa

### PR DESCRIPTION
**Pull request by JackBOBO**
1. add http xa pattern #68

### CODE REVIEW

File: src/Dtmcli/DtmClient.cs [hunk 1]
```diff
@@ -91,13 +90,10 @@ public async Task TransRegisterBranch(TransBase tb, Dictionary<string, string> a
...
         {
-            var queryParams = string.Format(
-                QueryStringFormat,
-                string.Concat(_dtmOptions.DtmUrl.TrimEnd(Slash), Constant.Request.URLBASE_PREFIX),
-                tb.Gid,
-                tb.TransType,
-                branchID,
-                op);
+            var uriPath = string.Concat(_dtmOptions.DtmUrl.TrimEnd(Slash), Constant.Request.URLBASE_PREFIX);
+            var queryParams = tb.TransType == "xa" ?
```
Comment (catcherwong) [inline src/Dtmcli/DtmClient.cs:94]:
```suggestion
            var queryParams = tb.TransType == DtmCommon.Constant.TYPE_XA ?
```

File: src/Dtmcli/Xa/XaLocalTransaction.cs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                var (_, ex) = await this._utils.InsertBarrier(conn, "xa", xa.Gid, xa.BranchIDGen.BranchID, DtmCommon.Constant.OpAction, xa.BranchIDGen.BranchID, xa.Op);
+                if (ex != null || conn.State != ConnectionState.Open)
+                    throw new DtmOngingException(ex?.Message);
+            }
+        }
+
+        private Xa XaFromQuery(IDictionary<string, string> values)
+        {
+            Xa xa = new(this._dtmClient);
+            xa.Gid = values[Request.GID];
```
Comment (catcherwong) [inline src/Dtmcli/Xa/XaLocalTransaction.cs:88]:
```suggestion
values.TryGetValue(Request.GID, out var gid)
```

File: src/Dtmcli/Xa/XaLocalTransaction.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+            if (DtmCommon.Constant.OpRollback == xa.Op)
+            {
+                var (_, ex) = await this._utils.InsertBarrier(conn, "xa", xa.Gid, xa.BranchIDGen.BranchID, DtmCommon.Constant.OpAction, xa.BranchIDGen.BranchID, xa.Op);
+                if (ex != null || conn.State != ConnectionState.Open)
+                    throw new DtmOngingException(ex?.Message);
+            }
+        }
+
+        private Xa XaFromQuery(IDictionary<string, string> values)
```
Comment (catcherwong) [inline src/Dtmcli/Xa/XaLocalTransaction.cs:85]: What about adding a new method to read the values from `IQueryCollection` ?

```cs
 private Xa XaFromQuery(Microsoft.AspNetCore.Http.IQueryCollection query)
```

And **XaFromQuery** should be puiblic?

File: src/Dtmcli/Xa/XaLocalTransaction.cs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        {
+            Xa xa = new(this._dtmClient);
+            xa.Gid = values[Request.GID];
+            xa.TransType = values[Request.TRANS_TYPE];
+            xa.Dtm = values.ContainsKey(Request.DTM) ? values[Request.DTM] : string.Empty;
+            xa.BranchIDGen = new BranchIDGen(values[Request.BRANCH_ID]);
+            xa.Op = values[Request.OP];
+            xa.Phase2Url = values.ContainsKey("phase2_url") ? values["phase2_url"] : string.Empty;
+            return xa;
```
Comment (catcherwong) [inline src/Dtmcli/Xa/XaLocalTransaction.cs:94]: Should check xa info here, valid or invalid.

File: src/Dtmcli/Xa/XaLocalTransaction.cs [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            this._logger = factory.CreateLogger<XaGlobalTransaction>();
+        }
+
+        public async Task Excecute(IDictionary<string, string> values, DbConnection conn, string dbType, Func<DbConnection, Xa, Task> xaFunc, CancellationToken token = default)
+        {
+            var xa = this.XaFromQuery(values);
+            var dbSpecial = this._dbSpecia.GetDbSpecialByName(dbType);
+            if (ConnectionState.Open != conn.State)
+            {
+                conn.Open();
```
Comment (catcherwong) [inline src/Dtmcli/Xa/XaLocalTransaction.cs:35]:
```suggestion
               await conn.OpenAsync();
```

**Review by catcherwong — CHANGES_REQUESTED** [review]

File: src/Dtmcli/Xa/XaLocalTransaction.cs (hunk 3 above)
Comment (JackBOBO) [reply src/Dtmcli/Xa/XaLocalTransaction.cs:85]: OK, I'll move "XaFromQuery" to "Xa" as a public static function and add an overload of IQueryCollection.

**Review by catcherwong — APPROVED** [review]

File: src/Dtmcli/Xa/Xa.cs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+            return await response.Content.ReadAsStringAsync();
+        }
+
+        [JsonIgnore]
+        public string Phase2Url { get; set; }
+
+
+#if NET5_0_OR_GREATER
+        public static Xa FromQuery(IDtmClient dtmClient, Microsoft.AspNetCore.Http.IQueryCollection quersy)
```
Comment (yedf2) [inline src/Dtmcli/Xa/Xa.cs:53]: `quersy` should be `query`

**Comment by yedf2** [comment]
For mysql version less than 8.0, connection should be closed after `xa end`, or else `xa commit` in another connection will not succeed. It is a bug of mysql, and is fixed in mysql 8

**Review by yedf2 — CHANGES_REQUESTED** [review]

**Comment by JackBOBO** [comment]
> For mysql version less than 8.0, connection should be closed after `xa end`, or else `xa commit` in another connection will not succeed. It is a bug of mysql, and is fixed in mysql 8

There is indeed this issue in versions below [MySQL 8](https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-29.html#mysqld-8-0-29-xa), and even after closing the connection after `xa end` or `xa commit`, the problem still persists. However, when I add `pooling=false` to the connection string, it seems to work fine. I suspect that the MysqlConnector connection pool cannot support this scenario.

**Comment by yedf2** [comment]
So for XA transaction, you may disable pool.

In the document of dtm, the performance of XA transaction are not high, so disable pooling should be OK

**Comment by catcherwong** [comment]
Can we merge this PR and release a new version now?

**Comment by JackBOBO** [comment]
> Can we merge this PR and release a new version now?

Of course you can.

**Pull request merged**